### **Context**
The movie rating system that film buffs know today has been around for over 50 years. Over the years, the cultural standards and norms have changed and so have movie ratings. However, even today the process of rating a film remains a closely guarded industry secret.


### **Objective**
MovieLens is a company in the internet and entertainment domain that has an online database of information related to films, television series, online streaming content – including cast, production crew, trivia, ratings, and fan and critical reviews. You have been hired as a Data Scientist for the company. You have been provided with the following three datasets, asked to carry out a detailed analysis of the data, and come up with some meaningful insights which will help the company to address their users in a better way.<br>
<br>
- **movie.csv**: The file contains information related to the movies and their genres. Columns: movie id, movie title, release date, Action, Adventure, Animation, Children’s, Comedy, Crime, Documentary, Drama, Fantasy, Film-Noir, Horror, Musical, Mystery, Romance, Sci-Fi, Thriller, War, Western<br>
- **user.csv**: It contains information about the users who have rated the movies. Columns: user id, age, gender, occupation, zip code<br>
- **ratings.csv**: It contains information for ratings given by the users to a particular movie. Columns: user id, movie id, rating, timestamp<br>
<br>

One of the first steps to carry out any analysis is to import the necessary libraries that will help us to carry out various operations on the data. <br><br>**NumPy, Pandas** are the most widely used python libraries in data science. It provides high-performance, easy to use structures and data analysis tools. <br>So let us first import **NumPy and Pandas** so that we can utilize the functions available in these libraries to analyze our data better.

### **Let us start by importing the required libraries**

In [ ]:
# Installing the libraries with the specified version.
!pip install numpy==1.25.2 pandas==1.5.3 -q --user

**Note**: *After running the above cell, kindly restart the notebook kernel and run all cells sequentially from the start again.*

In [ ]:
import pandas as pd
import numpy as np

Now we have imported pandas as pd and numpy as np. Here 'as' is used as an alias.

Let's load all the three datasets using Pandas **read** function so that we can start with our analysis using them

### **Reading the datasets**

In [ ]:
# Reading datasets by using read_csv from pandas package
movie = pd.read_csv("/content/movie.csv")
user = pd.read_csv("/content/user.csv")
ratings = pd.read_csv("/content/ratings.csv")

### **Data Overview**

The initial steps to understand any dataset is to:
- observe the first few rows of the dataset, to check whether the dataset has been loaded properly or not
- get information about the number of rows and columns in the dataset
- find out the data types of the columns to ensure that data is stored in the preferred format and the value of each property is as expected.
- check the statistical summary of the dataset to get an overview of the numerical columns of the data

Let us now delve deeper into each of the datasets individually and understand this aspects well

#### **User Dataset**

#### **Displaying the first 5 rows of the user dataset**

The **head()** function will help us to have a glimpse of the first 5 rows of the dataset

In [ ]:
user.head()

user id  age gender  occupation zip code
0        1   24      M  technician    85711
1        2   53      F       other    94043
2        3   23      M      writer    32067
3        4   24      M  technician    43537
4        5   33      F       other    15213

The user dataset has been loaded properly.

- *The user dataset contains demographic information about the users who have rated the movies*
- *The column `gender` has values as M and F, where M represents Male users and F represents Female users*
- *We can also observe that users are from different occupational backgrounds like 'technician', 'writer', etc.*
- *The column `zip code` represents the area where the user is residing*

#### **Understanding the shape of the user dataset**

Let us now check the dimensions of the dataset by using the ***shape*** attribute. <br>

The ***shape*** attribute will help us to get the number of rows and columns of the dataset.<br>

**Knowing the dimensions of the dataframe will help us to understand the number of data points we are working with and proves to be beneficial especially when the dataset is a large one (consisting of a large number of rows and columns).**

In [ ]:
user.shape

(943, 5)

- *There are 943 rows and 5 columns in the user dataset.*

#### **Checking the data types of the columns for the user dataset**

Now we are aware of the shape of the dataframe. Let us now have a look at the types of data that we are working with.<br>

The ***info()*** function will help us to understand the data types of the columns.<br>

**Understanding data types ensures that data is stored in the preferred format and the value of each property is as expected.**

In [ ]:
user.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 943 entries, 0 to 942
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   user id     943 non-null    int64 
 1   age         943 non-null    int64 
 2   gender      943 non-null    object
 3   occupation  943 non-null    object
 4   zip code    943 non-null    object
dtypes: int64(2), object(3)
memory usage: 37.0+ KB


- *We can see that the `gender`, `occupation` and `zip code` columns are of object type while the other columns are of integer type since they are containing numbers in them.*
- *The data type of the `zip code` column is object which seems to be wrong and needs further investigation since this column seems to be containing integer values only, as seen from the first 5 rows of the dataframe.*

#### **Getting the statistical summary for the user dataset**

We can now have a look at the statistical summary of the dataset.<br>

The **describe()** function will help us in finding out the statistical summary.<br>

**The statistical summary will give us a overview of the numerical columns of the data and display various features like the minimum, maximum, mean, standard deviation, etc. of the columns. This will help us to understand the distribution of the values present in the numerical columns.**

In [ ]:
user.describe()

user id         age
count  943.000000  943.000000
mean   472.000000   34.051962
std    272.364951   12.192740
min      1.000000    7.000000
25%    236.500000   25.000000
50%    472.000000   31.000000
75%    707.500000   43.000000
max    943.000000   73.000000

- *We can see from the `age` column that the average age of users who have rated the movies is 34 years*
- *The age of users ranges from 7 years (min or minimum) to 73 years (max or maximum)*
- *Since user id is an unique identifier of the users who have rated the movies, interpreting its summary statistics won't yield meaningful insights.*

Now we have somewhat got a basic idea of the user dataset. Let us now carry out similar operations on the other two datasets and try to find out relevant information from them as well.

#### **Movie Dataset**

#### **Displaying the first 5 rows of the movie dataset**

In [ ]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

- *As we can see the 'movie' dataset contains information related to the movies and their genres. While `movie id`, `movie title`, and `release date` are information related to the movies, the rest of the columns are specifically the genres of the movies.*<br>
- *The columns associated to the genres of the movies are having values of 0s and 1s. The value '1' means that a particular movie belongs to that genre while the value '0' represents that the movie is not a part of that genre.*
- *We can also see that a movie can have multiple genres as there are more than one column that are having value as '1' for a particular movie in the columns associated with genres.*<br>
- *For example, the movie named 'Toy Story' is having genres as 'Animation', 'Childrens', and 'Comedy'.*

#### **Understanding the shape of the movie dataset**

In [ ]:
movie.shape

(1680, 21)

- *There are 1680 rows and 21 columns in the movie dataset.*

#### **Checking the data types of the columns for the movie dataset**

In [ ]:
movie.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1680 entries, 0 to 1679
Data columns (total 21 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   movie id      1680 non-null   int64 
 1   movie title   1680 non-null   object
 2   release date  1680 non-null   object
 3   Action        1680 non-null   int64 
 4   Adventure     1680 non-null   int64 
 5   Animation     1680 non-null   int64 
 6   Childrens     1680 non-null   int64 
 7   Comedy        1680 non-null   int64 
 8   Crime         1680 non-null   int64 
 9   Documentary   1680 non-null   int64 
 10  Drama         1680 non-null   int64 
 11  Fantasy       1680 non-null   int64 
 12  Film-Noir     1680 non-null   int64 
 13  Horror        1680 non-null   int64 
 14  Musical       1680 non-null   int64 
 15  Mystery       1680 non-null   int64 
 16  Romance       1680 non-null   int64 
 17  Sci-Fi        1680 non-null   int64 
 18  Thriller      1680 non-null   int64 
 19  War   

- *We can see that the `movie title` and `release date` columns are of object type since these columns are having some text in them.*
- *All the other columns are of integer type since they are containing numbers in them.*

#### **Getting the statistical summary for the movie dataset**

In [ ]:
movie.describe()

movie id       Action    Adventure    Animation    Childrens  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean    841.525595     0.149405     0.080357     0.025000     0.072619   
std     485.609591     0.356593     0.271926     0.156171     0.259587   
min       1.000000     0.000000     0.000000     0.000000     0.000000   
25%     421.750000     0.000000     0.000000     0.000000     0.000000   
50%     841.500000     0.000000     0.000000     0.000000     0.000000   
75%    1261.250000     0.000000     0.000000     0.000000     0.000000   
max    1682.000000     1.000000     1.000000     1.000000     1.000000   

            Comedy        Crime  Documentary        Drama      Fantasy  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.300595     0.064881     0.029762     0.431548     0.013095   
std       0.458653     0.246389     0.169980     0.495440     0.113717   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       1.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

         Film-Noir       Horror      Musical      Mystery      Romance  \
count  1680.000000  1680.000000  1680.000000  1680.000000  1680.000000   
mean      0.014286     0.054762     0.033333     0.036310     0.147024   
std       0.118701     0.227583     0.179559     0.187115     0.354235   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

            Sci-Fi     Thriller          War      Western  
count  1680.000000  1680.000000  1680.000000  1680.000000  
mean      0.060119     0.149405     0.042262     0.016071  
std       0.237778     0.356593     0.201246     0.125788  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000

- *The statistical summary of the `movie_id` column will not yield meaningful insights as it is a unique identifier of each of the movies. The statistical summary of the columns related to genres is also not much helpful since these columns consists of only 0s and 1s values.*

We have now explored the movie dataset as well. Let us now finally see the ratings dataframe

#### **Ratings**

#### **Displaying the first 5 rows of the ratings dataset**

In [ ]:
ratings.head()

user id  movie id  rating  timestamp
0      196       242       3  881250949
1      186       302       3  891717742
2       22       377       1  878887116
3      244        51       2  880606923
4      166       346       1  886397596

The ratings dataset has also been imported properly without any errors

- *The ratings dataframe contains information about the ratings given by the users to a particular movie.*

#### **Understanding the shape of the ratings dataset**

In [ ]:
ratings.shape

(100000, 4)

- *There are 100000 rows and 4 columns in the ratings dataset. This is quite a large dataset as compared to the previous two dataframes.*

#### **Checking the data types of the columns for the ratings dataset**

In [ ]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 4 columns):
 #   Column     Non-Null Count   Dtype
---  ------     --------------   -----
 0   user id    100000 non-null  int64
 1   movie id   100000 non-null  int64
 2   rating     100000 non-null  int64
 3   timestamp  100000 non-null  int64
dtypes: int64(4)
memory usage: 3.1 MB


- *Here all the columns are of integer data type since they contain numbers in them.*

#### **Getting the statistical summary for the ratings dataset**

In [ ]:
ratings.describe()

user id       movie id         rating     timestamp
count  100000.00000  100000.000000  100000.000000  1.000000e+05
mean      462.48475     425.530130       3.529860  8.835289e+08
std       266.61442     330.798356       1.125674  5.343856e+06
min         1.00000       1.000000       1.000000  8.747247e+08
25%       254.00000     175.000000       3.000000  8.794487e+08
50%       447.00000     322.000000       4.000000  8.828269e+08
75%       682.00000     631.000000       4.000000  8.882600e+08
max       943.00000    1682.000000       5.000000  8.932866e+08

- *The numbers for the `timestamp` column do not convey any meaningful insights. However, from the `rating` column, we can see that the average rating of all the movies is around 3.53 while  the ratings ranges from 1 to 5.*

By now we have considerable idea about the datasets we are working with. It's time to dig deeper and find out answers to some important questions that will help us get a better idea about the industry.<br>


***Let's extract some meaningful insights from the data by answering a few business-oriented questions***

<h1><font size=6>How many movies belong to a particular genre?</font></h1>

We had seen earlier that a particular movie can have multiple genres. So let us investigate on this more.<br>

**But do you remember which dataset contains information about genres of movies ???**

*Yes*, movie dataset does contains this information. So let's try to have look on that

In [ ]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Fantasy  Film-Noir  Horror  \
0          1       1      0            0  ...        0          0       0   
1          0       0      0            0  ...        0          0       0   
2          0       0      0            0  ...        0          0       0   
3          0       1      0            0  ...        0          0       0   
4          0       0      1            0  ...        0          0       0   

   Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0        0        0        0       0         0    0        0  
1        0        0        0       0         1    0        0  
2        0        0        0       0         1    0        0  
3        0        0        0       0         0    0        0  
4        0        0        0       0         1    0        0  

[5 rows x 21 columns]

Let's analyze this dataset using some Pandas Functions

### **Pandas Functions**

#### **Getting all the columns of the movie dataset**

Let us print all the columns of the movie dataset so that we can get a better idea about what all genres are there

In [ ]:
movie.columns

Index(['movie id', 'movie title', 'release date', 'Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western'],
      dtype='object')

- *We can see that all the columns starting from 'Action' to 'Western' are columns that are related to the genres of the movies.*

Since there are many genres present in this dataset, let's create a variable to store all these values as a list so that we can use it for further analysis.

In [ ]:
genres = ['Action', 'Adventure',
       'Animation', 'Childrens', 'Comedy', 'Crime', 'Documentary', 'Drama',
       'Fantasy', 'Film-Noir', 'Horror', 'Musical', 'Mystery', 'Romance',
       'Sci-Fi', 'Thriller', 'War', 'Western']

In [ ]:
# using the len() function to calculate the number of genres
print('The number of genres present in the movie dataset are ', len(genres))

The number of genres present in the movie dataset are  18


- ***There are 18 genres in total.***

One of the way to find the answer is to take all the genre columns and find the sum for each column. Here, instead of writing the genre columns again, we will use the 'genres' list.

In [ ]:
# printing the genres list
genres

['Action',
 'Adventure',
 'Animation',
 'Childrens',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western']

In [ ]:
movie[genres].sum()

Action         251
Adventure      135
Animation       42
Childrens      122
Comedy         505
Crime          109
Documentary     50
Drama          725
Fantasy         22
Film-Noir       24
Horror          92
Musical         56
Mystery         61
Romance        247
Sci-Fi         101
Thriller       251
War             71
Western         27
dtype: int64

- *Here, we can see that 'Action' genre has 251 movies, 'Adventure' genre has 135 movies and so on. However it would have been far better if we had a series where the numbers were arranged in ascending or descending order so that we could have easily found out the genres with the highest and lowest number of movies.* <br>

This can be achieved by simply using the **sort_values()** function to arrange the numbers in descending or ascending order.

**sort_values() Function**<br>
Pandas sort_values() function sorts a data frame in Ascending or Descending order of passed Column. It’s different than the sorted Python function since it cannot sort a data frame and particular column cannot be selected.

In [ ]:
movie[genres].sum().sort_values(ascending = False) #ascending= False will sort the values in descending order. By default the parameter ascending is set to True

Drama          725
Comedy         505
Action         251
Thriller       251
Romance        247
Adventure      135
Childrens      122
Crime          109
Sci-Fi         101
Horror          92
War             71
Mystery         61
Musical         56
Documentary     50
Animation       42
Western         27
Film-Noir       24
Fantasy         22
dtype: int64

<h1><font size=5>Insights</font></h1>


- *Movies are distributed across 18 genres in total*
- *We can observe that 'Drama' genre has the highest number of movies with 725 in total followed by 'Comedy' at 505*
- *The top five genres with the highest number of movies are 'Drama', 'Comedy', 'Action', 'Thriller', and 'Romance'*
- *'Fantasy' genre has the lowest number of movies (22 in count)*

<h1><font size=6>Can we find the movies that have more than one genre?</font></h1>

As discussed earlier a particular movie can have more than one genre, so to answer the above problem we can use the below steps:


- The basic approach to find out this information will be to add a new column to the movie dataframe that will contain the summation of 1's from the genre columns along the row for a particular movie
- After this we can find out the number of movies that have more than one genre by simply picking up the rows where the added new column is having a value of more than 1

So, let us create a new column named `Number of Genres` and add it to the dataframe.

In [ ]:
movie["Number of Genres"] = movie.loc[:, genres].sum(axis=1)

- *The above code adds a new column named `Number of Genres` to the dataframe. This new column consists the summation of 1s from the genre columns along the row for a particular movie*<br>
- *movie.loc[:, genres] - This is picking up all the rows and the columns that are present in the genres list*
- *The axis = 1 parameter is used to get the summation of the values along the row while axis = 0 is used to get the summation of the values along the column*

Now let us print the head of the movie column and see whether the new column has been added or not.

In [ ]:
movie.head()

movie id  movie title release date  Action  Adventure  Animation  \
0         1   Toy Story      1-Jan-95       0          0          1   
1         2   GoldenEye      1-Jan-95       1          1          0   
2         3  Four Rooms      1-Jan-95       0          0          0   
3         4  Get Shorty      1-Jan-95       1          0          0   
4         5     Copycat      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  Film-Noir  Horror  Musical  \
0          1       1      0            0  ...          0       0        0   
1          0       0      0            0  ...          0       0        0   
2          0       0      0            0  ...          0       0        0   
3          0       1      0            0  ...          0       0        0   
4          0       0      1            0  ...          0       0        0   

   Mystery  Romance  Sci-Fi  Thriller  War  Western  Number of Genres  
0        0        0       0         0    0        0                 3  
1        0        0       0         1    0        0                 3  
2        0        0       0         1    0        0                 1  
3        0        0       0         0    0        0                 3  
4        0        0       0         1    0        0                 3  

[5 rows x 22 columns]

- *As we can see a new column named `Number of Genres` has been added at the end of the dataframe*

Now let us pick up the rows where `Number of Genres`>1 so that we can find the number of movies with more than 1 genre

In [ ]:
movie[movie['Number of Genres']>1] #this will subset the data as per the given condition

movie id         movie title release date  Action  Adventure  Animation  \
0            1          Toy Story      1-Jan-95       0          0          1   
1            2          GoldenEye      1-Jan-95       1          1          0   
3            4         Get Shorty      1-Jan-95       1          0          0   
4            5            Copycat      1-Jan-95       0          0          0   
6            7     Twelve Monkeys      1-Jan-95       0          0          0   
...        ...                 ...          ...     ...        ...        ...   
1666      1669  MURDER and murder     20-Jun-97       0          0          0   
1667      1670            Tainted      1-Feb-98       0          0          0   
1670      1673             Mirage      1-Jan-95       1          0          0   
1676      1679          B. Monkey      6-Feb-98       0          0          0   
1677      1680      Sliding Doors      1-Jan-98       0          0          0   

      Childrens  Comedy  Crime  Documentary  ...  Film-Noir  Horror  Musical  \
0             1       1      0            0  ...          0       0        0   
1             0       0      0            0  ...          0       0        0   
3             0       1      0            0  ...          0       0        0   
4             0       0      1            0  ...          0       0        0   
6             0       0      0            0  ...          0       0        0   
...         ...     ...    ...          ...  ...        ...     ...      ...   
1666          0       0      1            0  ...          0       0        0   
1667          0       1      0            0  ...          0       0        0   
1670          0       0      0            0  ...          0       0        0   
1676          0       0      0            0  ...          0       0        0   
1677          0       0      0            0  ...          0       0        0   

      Mystery  Romance  Sci-Fi  Thriller  War  Western  Number of Genres  
0           0        0       0         0    0        0                 3  
1           0        0       0         1    0        0                 3  
3           0        0       0         0    0        0                 3  
4           0        0       0         1    0        0                 3  
6           0        0       1         0    0        0                 2  
...       ...      ...     ...       ...  ...      ...               ...  
1666        1        0       0         0    0        0                 3  
1667        0        0       0         1    0        0                 2  
1670        0        0       0         1    0        0                 2  
1676        0        1       0         1    0        0                 2  
1677        0        1       0         0    0        0                 2  

[849 rows x 22 columns]

Let us print the shape of this dataframe so that we know the number of movies that are having more than one genre.

In [ ]:
movie[movie['Number of Genres']>1].shape

(849, 22)

- *There are 849 movies that are having more than one genre.*

In [ ]:
movie.shape

(1680, 22)

- *The original movie dataset contained 1680 movies. So out of 1680 movies, 849 i.e around 50% of the movies are having more than one genre.*

<h1><font size=5>Insights</font></h1>


- *Out of 1680 movies, 849 movies comprising around 50% are having more than one genre and 831 movies have only single genre*
- *This shows that we have an almost equal ratio (50:50) of movies having more than one genre and having single genre*

<h1><font size=6>Can we find the genres that are most liked by the users?</font></h1>


*As we have already seen that the information regarding the ratings is present in the ratings dataset, however, the genre columns are present in the movie dataset. The only column connecting this two datasets is the movie id column. We can `merge` both the datasets on the basis of common variable that is `movie_id`.*

**Merge/Join / Concatenation**<br>
Pandas provide various facilities for easily combining together Series or DataFrame with various kinds of set logic for the indexes and relational algebra functionality in the case of join / merge-type operations.<br> <br>Here are the different types of the Joins:<br>

**Inner Join**: Returns records that have matching values in both tables<br>

**Left Join**: Returns all records from the left table, and the matched records from the right table<br>

**Right Join**: Returns all records from the right table, and the matched records from the left table<br>

**Outer/Full Join**: Returns all records when there is a match in either left or right table

In [ ]:
df_merge = movie.merge(ratings, on = 'movie id', how = 'inner')
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Mystery  Romance  Sci-Fi  Thriller  War  \
0       1      0            0  ...        0        0       0         0    0   
1       1      0            0  ...        0        0       0         0    0   
2       1      0            0  ...        0        0       0         0    0   
3       1      0            0  ...        0        0       0         0    0   
4       1      0            0  ...        0        0       0         0    0   

   Western  Number of Genres  user id  rating  timestamp  
0        0                 3      308       4  887736532  
1        0                 3      287       5  875334088  
2        0                 3      148       4  877019411  
3        0                 3      280       4  891700426  
4        0                 3       66       3  883601324  

[5 rows x 25 columns]

- *As we can see that the `user id`, `rating`, `timestamp` columns are added at the end of the dataframe*

In order to find the genres which are most liked by the users, we can find the average ratings of each of the genre and then compare among them.
<br>
For this we can use the `genres` list

In [ ]:
genres

['Action',
 'Adventure',
 'Animation',
 'Childrens',
 'Comedy',
 'Crime',
 'Documentary',
 'Drama',
 'Fantasy',
 'Film-Noir',
 'Horror',
 'Musical',
 'Mystery',
 'Romance',
 'Sci-Fi',
 'Thriller',
 'War',
 'Western']

We will run a loop across the merged dataframe for the columns that are present in the genres list. Whenever a particular genre is found to be have '1' value in its column, the rating corresponding to that '1' value will be extracted and in this end we will calculate the average value of all this extracted ratings for a particular genre.

In [ ]:
for i in genres:
    print(i,':' , df_merge[df_merge[i]==1].rating.mean())

Action : 3.480245417953027
Adventure : 3.503526503308369
Animation : 3.5766990291262135
Childrens : 3.3532442216652742
Comedy : 3.3940734781442745
Crime : 3.6322780881440098
Documentary : 3.6728232189973613
Drama : 3.6873793708484772
Fantasy : 3.2152366863905324
Film-Noir : 3.9215233698788228
Horror : 3.2903893172841827
Musical : 3.521396851029471
Mystery : 3.63813155386082
Romance : 3.621704948358255
Sci-Fi : 3.5607227022780834
Thriller : 3.5090069495245064
War : 3.815811874866993
Western : 3.6132686084142396


<h1><font size=5>Insights</font></h1>


- *'Film-Noir' genre is having the highest average ratings with an average rating of 3.92 followed by 'War' genre with rating of 3.81*
- *'Fantasy' genre is having the lowest average ratings with an average rating of 3.21*
- *Among 18 genres, only 5 genres have received a rating below 3.5 which means around 72% of genres are rated above average*

Till now we have done analyses on the genres and movies and generated some insights.

<h1><font size=6>Can we find which movies have been most preferred by the users?</font></h1>

Let's start by checking the top rated movies by the users.
As we already know that `movie` and `ratings` datasets has the information related to the ratings of the movies. <br>
Therefore, we can *merge* both the datasets on the basis of common variable that is `movie_id` and get the average ratings for all the movies. Since we have already merged these datasets before, we can call it again to get the values

In [ ]:
df_merge.head()

movie id movie title release date  Action  Adventure  Animation  Childrens  \
0         1  Toy Story      1-Jan-95       0          0          1          1   
1         1  Toy Story      1-Jan-95       0          0          1          1   
2         1  Toy Story      1-Jan-95       0          0          1          1   
3         1  Toy Story      1-Jan-95       0          0          1          1   
4         1  Toy Story      1-Jan-95       0          0          1          1   

   Comedy  Crime  Documentary  ...  Mystery  Romance  Sci-Fi  Thriller  War  \
0       1      0            0  ...        0        0       0         0    0   
1       1      0            0  ...        0        0       0         0    0   
2       1      0            0  ...        0        0       0         0    0   
3       1      0            0  ...        0        0       0         0    0   
4       1      0            0  ...        0        0       0         0    0   

   Western  Number of Genres  user id  rating  timestamp  
0        0                 3      308       4  887736532  
1        0                 3      287       5  875334088  
2        0                 3      148       4  877019411  
3        0                 3      280       4  891700426  
4        0                 3       66       3  883601324  

[5 rows x 25 columns]

- *The datasets has been merged successfully and the columns has been added at the end of the dataset*

We can notice that multiple rating has been given to a single movie from different users.

**To find out movies preferred by users we need to first get the average ratings of each movie** <br><br> We can perform this by using *group_by* function from Pandas

**Groupby Function**<br>
Pandas’ *GroupBy* is a powerful and versatile function in Python. It allows you to split your data into separate groups to perform computations for better analysis.*GroupBy* allows us to group our data based on different features and get a more accurate idea about your data.

Now let's apply the groupby function to our datasets:

In [ ]:
avg_rating = df_merge.groupby(['movie title'])[['rating']].mean().reset_index() # reset_index will reset the index of dataframe to default indexing (0 to number of rows minus 1)

avg_rating

movie title    rating
0                 'Til There Was You   2.333333
1                              1-900   2.600000
2                     101 Dalmatians   2.908257
3                       12 Angry Men   4.344000
4                                 187  3.024390
...                               ...       ...
1652                      Young Guns   3.207921
1653                   Young Guns II   2.772727
1654  Young Poisoner's Handbook, The   3.341463
1655                Zeus and Roxanne   2.166667
1656     Á köldum klaka (Cold Fever)   3.000000

[1657 rows x 2 columns]

*This displays the average rating for all the movies.* <br><br>
Let's rename the `rating` column to `Avg_rating` for better understanding



In [ ]:
avg_rating.rename(columns={'rating':'Avg_rating'},inplace=True)
avg_rating.head()

movie title  Avg_rating
0  'Til There Was You     2.333333
1               1-900     2.600000
2      101 Dalmatians     2.908257
3        12 Angry Men     4.344000
4                  187    3.024390

We can also sort them in a descending or ascending order by using *sort_values* function
<br>**Let's check the movies with highest average ratings**

In [ ]:
avg_rating.sort_values(ascending=False, by= 'Avg_rating')

movie title  Avg_rating
631                       Great Day in Harlem, A          5.0
1170                                 Prefontaine          5.0
1354                      Someone Else's America          5.0
461   Entertaining Angels: The Dorothy Day Story          5.0
942          Marlene Dietrich: Shadow and Light (         5.0
...                                           ...         ...
111                                    Babyfever          1.0
834                               Lashou shentan          1.0
1307                            Shadows (Cienie)          1.0
1304       Shadow of Angels (Schatten der Engel)          1.0
1168                                    Power 98          1.0

[1657 rows x 2 columns]

In [ ]:
avg_rating[avg_rating['Avg_rating']==5] #it will subset the data as per the specified condition

movie title  Avg_rating
30                                 Aiqing wansui          5.0
461   Entertaining Angels: The Dorothy Day Story          5.0
631                       Great Day in Harlem, A          5.0
942          Marlene Dietrich: Shadow and Light (         5.0
1170                                 Prefontaine          5.0
1270               Saint of Fort Washington, The          5.0
1274                          Santa with Muscles          5.0
1354                      Someone Else's America          5.0
1382                                    Star Kid          5.0
1466                     They Made Me a Criminal          5.0

<h1><font size=5>Insights</font></h1>


- *We can observe movies like **Great Day in Harlem, A, Prefontaine** etc are top rated movies*
- *Movies like **Shadow of Angels (Schatten der Engel), Power 98**, etc are least rated*
- *There are 10 movies which have been rated as 5.0*

But we know there are movies which are rated more than once by different users, and hence this may affect the overall rating of the movie.

<h1><font size=6>So can we find which movies are rated most of the times?</font></h1>

Let's create a variable which contains the movies with their ratings count

In [ ]:
movie_count = df_merge.groupby(['movie title'])['rating'].count().reset_index()
movie_count

movie title  rating
0                 'Til There Was You        9
1                              1-900        5
2                     101 Dalmatians      109
3                       12 Angry Men      125
4                                 187      41
...                               ...     ...
1652                      Young Guns      101
1653                   Young Guns II       44
1654  Young Poisoner's Handbook, The       41
1655                Zeus and Roxanne        6
1656     Á köldum klaka (Cold Fever)        1

[1657 rows x 2 columns]

We can again rename the `rating` column to `Rating_counts` and sort them in descending order to display highest to lowest count of ratings

In [ ]:
movie_count.rename(columns={'rating':'Rating_counts'},inplace=True)
movie_count.head()

movie title  Rating_counts
0  'Til There Was You               9
1               1-900               5
2      101 Dalmatians             109
3        12 Angry Men             125
4                  187             41

In [ ]:
movie_count.sort_values(ascending=False, by= 'Rating_counts')

movie title  Rating_counts
1393                    Star Wars             583
332                       Contact             509
497                         Fargo             508
1232           Return of the Jedi             507
858                     Liar Liar             485
...                            ...            ...
1085  Object of My Affection, The               1
1088                Office Killer               1
1109    Other Voices, Other Rooms               1
1123            Paris Was a Woman               1
1656  Á köldum klaka (Cold Fever)               1

[1657 rows x 2 columns]

- *We can observe that **Star Wars** has been rated most number of times (**583 times**) and there are few movies which have least rating counts like **Paris Was a Woman	,Á köldum klaka (Cold Fever)**.*

There are other movies which are mostly rated by users, so let's check movies which are rated more than 100 times

In [ ]:
movie_100 = movie_count[movie_count['Rating_counts']>100] #this will subset the dataset 'movie_count' as per the specified condition
movie_100

movie title  Rating_counts
2                            101 Dalmatians             109
3                              12 Angry Men             125
7                     2001: A Space Odyssey             259
15                           Absolute Power             127
16                               Abyss, The             151
...                                      ...            ...
1627  Willy Wonka and the Chocolate Factory             326
1638                      Wizard of Oz, The             246
1646                    Wrong Trousers, The             118
1651                     Young Frankenstein             200
1652                             Young Guns             101

[334 rows x 2 columns]

*There are 334 movies which are rated more than 100 times by the users*

Among these movies, let's find out **which movies has been top rated by customers** so that we can know what kind of movies are preferred by users mostly

In [ ]:
avg_rating.head()

movie title  Avg_rating
0  'Til There Was You     2.333333
1               1-900     2.600000
2      101 Dalmatians     2.908257
3        12 Angry Men     4.344000
4                  187    3.024390

In [ ]:
movie_100.head()

movie title  Rating_counts
2          101 Dalmatians             109
3            12 Angry Men             125
7   2001: A Space Odyssey             259
15         Absolute Power             127
16             Abyss, The             151

We can merge both the datasets, and have a combined view on the basis of Average rating and rating counts

In [ ]:
df_top= avg_rating.merge(movie_100, on = 'movie title', how = 'inner')
df_top

movie title  Avg_rating  Rating_counts
0                           101 Dalmatians     2.908257            109
1                             12 Angry Men     4.344000            125
2                    2001: A Space Odyssey     3.969112            259
3                           Absolute Power     3.370079            127
4                               Abyss, The     3.589404            151
..                                      ...         ...            ...
329  Willy Wonka and the Chocolate Factory     3.631902            326
330                      Wizard of Oz, The     4.077236            246
331                    Wrong Trousers, The     4.466102            118
332                     Young Frankenstein     3.945000            200
333                             Young Guns     3.207921            101

[334 rows x 3 columns]

In [ ]:
df_top.sort_values(ascending=False, by='Avg_rating').head(25) #display top 25 movies on the basis of average rating

movie title  Avg_rating  \
63                                     Close Shave, A     4.491071   
260                                  Schindler's List     4.466443   
331                               Wrong Trousers, The     4.466102   
54                                         Casablanca     4.456790   
269                         Shawshank Redemption, The     4.445230   
248                                       Rear Window     4.387560   
319                               Usual Suspects, The     4.385768   
291                                         Star Wars     4.358491   
1                                        12 Angry Men     4.344000   
59                                       Citizen Kane     4.292929   
305                             To Kill a Mockingbird     4.292237   
222                   One Flew Over the Cuckoo's Nest     4.291667   
272                         Silence of the Lambs, The     4.289744   
218                                North by Northwest     4.284916   
131                                    Godfather, The     4.283293   
264                                    Secrets & Lies     4.265432   
135                                 Good Will Hunting     4.262626   
189                         Manchurian Candidate, The     4.259542   
91   Dr. Strangelove or: How I Learned to Stop Worr...    4.252577   
243                           Raiders of the Lost Ark     4.252381   
320                                           Vertigo     4.251397   
304                                           Titanic     4.245714   
176                                Lawrence of Arabia     4.231214   
188                               Maltese Falcon, The     4.210145   
98                           Empire Strikes Back, The     4.204360   

     Rating_counts  
63             112  
260            298  
331            118  
54             243  
269            283  
248            209  
319            267  
291            583  
1              125  
59             198  
305            219  
222            264  
272            390  
218            179  
131            413  
264            162  
135            198  
189            131  
91             194  
243            420  
320            179  
304            350  
176            173  
188            138  
98             367

Similarly we can sort them as per the ratings counts also

In [ ]:
df_top.sort_values(ascending=False, by='Rating_counts').head(25) #display top 25 movies on the basis of rating counts

movie title  Avg_rating  Rating_counts
291                           Star Wars     4.358491            583
69                              Contact     3.803536            509
107                               Fargo     4.155512            508
251                  Return of the Jedi     4.007890            507
178                           Liar Liar     3.156701            485
99                 English Patient, The     3.656965            481
261                              Scream     3.441423            478
309                           Toy Story     3.878319            452
8                         Air Force One     3.631090            431
159              Independence Day (ID4)     3.438228            429
243             Raiders of the Lost Ark     4.252381            420
131                      Godfather, The     4.283293            413
240                        Pulp Fiction     4.060914            394
314                      Twelve Monkeys     3.798469            392
272           Silence of the Lambs, The     4.289744            390
167                       Jerry Maguire     3.710938            384
55                          Chasing Amy     3.839050            379
254                           Rock, The     3.693122            378
98             Empire Strikes Back, The     4.204360            367
287            Star Trek: First Contact     3.660274            365
304                             Titanic     4.245714            350
25                   Back to the Future     3.834286            350
199                 Mission: Impossible     3.313953            344
119                       Fugitive, The     4.044643            336
160  Indiana Jones and the Last Crusade     3.930514            331

<h1><font size=5>Insights</font></h1>


- *There are 334 movies in total which have received ratings more than 100 times*
- *As per the average rating, **Close Shave, A** has maximum average rating with a rating count of 112*
- *As per the rating counts, **Star Wars** has received maximum number counts with an average rating of 4.35*
- *There are few movies which have least rating counts like Paris Was a Woman	,Á köldum klaka (Cold Fever)*

Since we have extracted the top rated movies with most number of ratings given by users.

<h1><font size=6>Can we extract some relationship between demographic details of the users and ratings for the movies?</font></h1>

****



All the datasets can be used to identify the demography of the users. Therefore, we will merge all the datasets together and perform the analysis

Since we have already merged datasets `rating` and `movie` in a variable `df_merge`, therefore we can merge the dataset `user` with this dataset

In [ ]:
# Merging user dataset with movie and ratings(already merged : df_merge) dataset
df_merge_all = df_merge.merge(user, on = 'user id', how = 'inner')

In [ ]:
df_merge_all.head()

movie id      movie title release date  Action  Adventure  Animation  \
0         1       Toy Story      1-Jan-95       0          0          1   
1         4      Get Shorty      1-Jan-95       1          0          0   
2         5         Copycat      1-Jan-95       0          0          0   
3         7  Twelve Monkeys      1-Jan-95       0          0          0   
4         8            Babe      1-Jan-95       0          0          0   

   Childrens  Comedy  Crime  Documentary  ...  War  Western  Number of Genres  \
0          1       1      0            0  ...    0        0                 3   
1          0       1      0            0  ...    0        0                 3   
2          0       0      1            0  ...    0        0                 3   
3          0       0      0            0  ...    0        0                 2   
4          1       1      0            0  ...    0        0                 3   

   user id  rating  timestamp  age  gender  occupation  zip code  
0      308       4  887736532   60       M     retired     95076  
1      308       5  887737890   60       M     retired     95076  
2      308       4  887739608   60       M     retired     95076  
3      308       4  887738847   60       M     retired     95076  
4      308       5  887736696   60       M     retired     95076  

[5 rows x 29 columns]

**Let's check on various attributes of the users in order to know the preferences for the movies and their genres**

In [ ]:
user['gender'].value_counts(normalize=True) #this will give the proportion of males and females users extracted from User dataset

M    0.710498
F    0.289502
Name: gender, dtype: float64

In [ ]:
df_merge_all.groupby('gender').rating.mean() #grouping the dataset on 'gender' column and taking the average of ratings

gender
F    3.531510
M    3.529333
Name: rating, dtype: float64

- *The average rating given by Males and Females is almost equal with a value of around 3.53*

In [ ]:
df_merge_all.groupby('occupation').rating.mean().sort_values(ascending= False) #grouping the dataset on 'occupation' column and taking the average of ratings.

occupation
none             3.777778
lawyer           3.735316
doctor           3.688889
educator         3.670692
artist           3.653380
administrator    3.635646
scientist        3.611273
salesman         3.582944
programmer       3.568205
librarian        3.560781
other            3.552335
engineer         3.541473
technician       3.532097
student          3.515167
marketing        3.485641
retired          3.466750
entertainment    3.440783
writer           3.376152
executive        3.349794
homemaker        3.301003
healthcare       2.896220
Name: rating, dtype: float64

- *Users who are not working have given highest ratings as compared to users who are working professionals*
- *Healthcare workers have given lowest ratings on an average*

In [ ]:
df_merge_all.groupby(['occupation','gender']).rating.mean()

occupation     gender
administrator  F         3.781839
               M         3.555233
artist         F         3.347065
               M         3.875841
doctor         M         3.688889
educator       F         3.699132
               M         3.660246
engineer       F         3.751724
               M         3.537676
entertainment  F         3.448889
               M         3.439807
executive      F         3.773756
               M         3.320340
healthcare     F         2.736021
               M         3.639839
homemaker      F         3.278810
               M         3.500000
lawyer         F         3.623188
               M         3.741379
librarian      F         3.580070
               M         3.537920
marketing      F         3.522624
               M         3.474801
none           F         3.632877
               M         3.876636
other          F         3.531114
               M         3.563447
programmer     F         3.577566
               M         3

<h1><font size=5>Insights</font></h1>


- *Around 71% of users are Males while only 29% users are females*
- *The average rating given by Males and Females is almost equal with a value of around 3.53*
- *Users who are not working have given highest ratings as compared to users who are working professionals*
- *Healthcare workers have given lowest ratings on an average*
- *The low ratings from the healthcare sector is majorly driven by female workers*
- *Only male doctors have given ratings. No female doctors have given ratings as per the observations*

### **Executive Summary**

- *Movies are distributed across 18 genres in total*
- *The top five genres with the highest number of movies are 'Drama', 'Comedy', 'Action', 'Thriller', and 'Romance'*
- *Out of 1680 movies, 50% of the movies have more than one genre, while the rest have a single genre*
- The *'Film-Noir' genre has the highest average ratings with an average rating of 3.92, while the 'Fantasy' genre has the lowest average ratings with an average rating of 3.21*
- *Among 18 genres, 72% of the genres are rated above average 3.5*
- *Movies like **Great Day in Harlem, A**, **Prefontaine**, etc. are top-rated movies, while movies like **Shadow of Angels (Schatten der Engel)**, **Power 98**, etc. are least rated*
- *There are 334 movies in total which have been rated more than 100 times*
- *The movie **Close Shave, A** has the highest average rating and was rated 112 times, while the **Star Wars** movie received the highest number of ratings with an average rating of 4.35*
- *Around 71% of users are Males while only 29% users are females*
- *The average rating given by Males and Females is almost equal with a value of around 3.53*
- *Users who are not working have rated movies higher compared to users who are working professionals*
- *Healthcare workers have given the lowest ratings on an average*

---